### PR TITLE
Add support for an initialization handshake

### LINKED ISSUE #22: Create a handshake to know that heart:set_callback/2 has been called
When using [`heart:set_callback/2`](https://www.erlang.org/doc/man/heart.html#set_callback-2), there's a race condition between startup and the code that calls that function. If, for whatever reason, the code path that registers the callback function isn't run, Erlang will happily report to `heart` that everything is ok even if that check would have said otherwise. 

One option is to call `heart:set_callback/2` as early as possible, but this doesn't completely fix the issue since there's still a gap. The proposal is for `nerves_heart` to require a notification that `set_callback/2` has been called for it to continue to pet the hardware watchdog and not reboot. The details for how this handshake occurs are TBD.

### PR DESCRIPTION
**Pull request by fhunleth**
When enabled, Nerves Heart will expect an initialization
handshake message in addition to the regular heart beat messages. This
makes it possible to call `:heart.set_callback/1` without worrying about
an error that causes the code that calls `:heart.set_callback/1` somehow
not getting run.

Fixes #22

### CODE REVIEW

File: src/heart.c [hunk 1]
```diff
@@ -160,6 +161,15 @@ static int wdt_timeout = 60;
 /* last_wdt_pet_time is the absolute time that hardware watchdog was pet */
 static time_t last_wdt_pet_time = 0;
 
+/* Timeout on receiving a handshake message from the application that the heart callback. 0=unused */
```
Comment (jjcarstens) [inline src/heart.c]:
```suggestion
/* Timeout on receiving a handshake message from the application that the heart callback was set. 0=unused */
```
I believe this is what you meant?

File: src/heart.c [hunk 2]
```diff
@@ -160,6 +161,15 @@ static int wdt_timeout = 60;
...
+static time_t init_handshake_timeout = 0;
+
+/* Set to 1 if the initialization handshake message came in */
+static int init_handshake_happened = 0;
+
+/* If !init_handshake_happened, then this it the end time */
```
Comment (jjcarstens) [inline src/heart.c]:
```suggestion
/* If !init_handshake_happened, then this is the end time */
```

File: src/heart.c [hunk 3]
```diff
@@ -348,6 +358,15 @@ int main(int argc, char **argv)
 
     print_log("heart: " PROGRAM_NAME " v" PROGRAM_VERSION_STR " started.");
 
+    // Assume that the handshake happened and this fix it if a timeout was specified
```
Comment (jjcarstens) [inline src/heart.c]:
```suggestion
    // Assume that the handshake happened and this fixes it if a timeout was specified
```

**Review by jjcarstens — COMMENTED** [review]
I've only gotten partially through this, but looking good so far

**Comment by fhunleth** [comment]
Thanks for all of the comment fixes. I accepted them all.

**Review by jjcarstens — APPROVED** [review]

**Pull request merged**